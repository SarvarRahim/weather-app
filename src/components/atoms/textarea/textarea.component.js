import React from 'react';
import { hoc } from '@utils/hoc';
import { useTextareaProps } from './textarea.props';
import {
  TextareaContainer,
  TextareaField,
  TextareaSuffixContainer
} from './textarea.style';

/**
 * <Textarea />
 */
export const Textarea = hoc(
  useTextareaProps,
  ({ suffix, _onChange, _textareaRef, suffixPadding, ...rest }) => (
    <TextareaContainer suffixPadding={suffixPadding}>
      <TextareaField
        data-testid='textarea-field'
        rows={1}
        {...rest}
        ref={_textareaRef}
        onChange={_onChange}
      />
      <TextareaSuffixContainer>{suffix}</TextareaSuffixContainer>
    </TextareaContainer>
  )
);
