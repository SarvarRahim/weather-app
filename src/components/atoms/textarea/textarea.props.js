import { useImperativeHandle, useRef } from 'react';

/**
 * <Textarea /> props
 */
export const useTextareaProps = ({ onChange, textareaRef }) => {
  const _textareaRef = useRef();

  const _onChange = e => {
    const value = e.target?.value;
    if (_textareaRef.current) {
      _textareaRef.current.style.height = 'auto';
      _textareaRef.current.style.height =
        _textareaRef.current.scrollHeight + 'px';
    }
    onChange?.(value);
  };

  useImperativeHandle(textareaRef, () => ({
    focus: () => {
      _textareaRef.current.focus();
    },
    get value() {
      return _textareaRef.current.value;
    },
    set value(val) {
      _textareaRef.current.value = val;
    }
  }));

  return {
    _onChange,
    _textareaRef
  };
};
