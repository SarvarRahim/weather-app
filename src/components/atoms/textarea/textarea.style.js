import styled from 'styled-components';

export const TextareaContainer = styled.div`
  position: relative;
  z-index: 1;
  flex-grow: 1;
  width: 100%;
  border: 2px solid #dadada;
  border-radius: 8px;
  box-sizing: border-box;
  padding: ${props => props.suffixPadding || '16px 68px 16px 24px'};
  display: flex;
  align-items: center;
`;

export const TextareaField = styled.textarea`
  width: ${props => props.width || '100%'};
  box-sizing: border-box;
  border-radius: ${props => props.rounded || '5px'};
  font-weight: 400;
  font-size: 16px;
  line-height: 20px;
  resize: none;
  border: none;
  outline: none;

  &::-webkit-scrollbar {
    display: none;
  }

  &::placeholder {
    font-weight: 400;
    font-size: ${props => props.fontSize || '16px'};
    color: #c4c4c4;
  }
`;

export const TextareaSuffixContainer = styled.div`
  position: absolute;
  z-index: 2;
  top: ${props => props.top || 'calc(100% / 2 - 24px)'};
  bottom: ${props => props.top || 'calc(100% / 2 - 24px)'};
  margin: auto;
  right: ${props => props.right || '14px'};
  height: max-content;
`;
