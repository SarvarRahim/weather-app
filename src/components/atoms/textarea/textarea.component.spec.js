import { fireEvent, screen } from '@testing-library/react';
import { customRender } from '@test-utils';
import { Textarea } from './textarea.component';
import 'jest-styled-components';
import '@testing-library/jest-dom';

const setup = ({ width, rounded, placeholder, onChange }) => {
  return customRender(
    <Textarea
      width={width}
      rounded={rounded}
      placeholder={placeholder}
      onChange={onChange}
    />
  );
};

const handleChange = jest.fn();
const MESSAGE = 'test message';

const EVENT = { target: { value: MESSAGE } };

test('Render Textarea component', () => {
  const { asFragment } = setup({
    width: '100%',
    rounded: '5px',
    placeholder: 'placeholder'
  });
  expect(asFragment).toMatchSnapshot();
});

test('Render width', () => {
  setup({
    width: '100%'
  });
  expect(screen.getByTestId('textarea-field')).toHaveStyle('width: 100%');
});

test('Render rounded', () => {
  setup({
    width: '100%',
    rounded: '5px'
  });
  expect(screen.getByTestId('textarea-field')).toHaveStyle(
    'width: 100%',
    'rounded: 5px'
  );
});

test('Render placeholder', () => {
  setup({
    placeholder: 'placeholder'
  });
  expect(screen.getByPlaceholderText(/placeholder/i)).toBeInTheDocument();
});

test('Should change value', () => {
  setup({
    onChange: handleChange
  });
  const textarea = screen.getByTestId('textarea-field');
  expect(textarea.value).toBe('');
  fireEvent.change(textarea, EVENT);
  expect(handleChange).toHaveBeenCalledWith(MESSAGE);
});
