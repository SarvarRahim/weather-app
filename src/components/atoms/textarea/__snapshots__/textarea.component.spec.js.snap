// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Textarea component 1`] = `[Function]`;
